lib/identity/hostdata/config_reader.rb

```diff
@@ -0,0 +1,106 @@
+module Identity
+  module Hostdata
+    class ConfigReader
+      attr_reader :app_root, :logger
+
+      def initialize(
+        app_root: Rails.root,
+        s3_client: nil,
+        logger: Logger.new(STDOUT)
+      )
+        @app_root = app_root
+        @s3_client = s3_client
+        @logger = logger
+      end
+
+      def read_configuration(rails_env, write_copy_to: nil)
+        if write_copy_to && !File.exists?(write_copy_to)
+          FileUtils.mkdir_p(File.dirname(write_copy_to))
+          File.write(write_copy_to, base_configuration.to_yaml)
+          FileUtils.chmod(0o640, write_copy_to)
+        end
+
+        base_configs = base_configuration.dup
+        base_configs.delete('development')
+        base_configs.delete('production')
+        base_configs.delete('test')
+        base_configs.merge(
+          base_configuration[rails_env],
+        ).transform_keys(&:to_sym)
+      end
+
+      private
+
+      def base_configuration
+        @base_configuration ||= deep_merge(
+          deep_merge(default_configuration, app_override_configuration),
+          role_override_configuration,
+        )
+      end
+
+      def default_configuration
+        YAML.safe_load(File.read(File.join(app_root, 'config', 'application.yml.default')))
+      end
+
+      def app_override_configuration
+        local_config_filepath = File.join(app_root, 'config', 'application.yml')
+        raw_configs = if Identity::Hostdata.in_datacenter?
+                        app_secrets_s3.read_file(app_configuration_s3_path)
+                      elsif File.exists?(local_config_filepath)
+                        File.read(local_config_filepath)
+                      end
+        YAML.safe_load(raw_configs || '{}') || {}
+      end
+
+      def role_override_configuration
+        return {} if role_configuration_filename.nil?
+        local_config_filepath = File.join(app_root, 'config', role_configuration_filename)
+
+        raw_configs = if Identity::Hostdata.in_datacenter?
+                        app_secrets_s3.read_file(role_configuration_s3_path)
+                      elsif File.exists?(local_config_filepath)
+                        File.read(local_config_filepath)
+                      end
+
+        YAML.safe_load(raw_configs || '{}') || {}
+      end
+
+      def app_secrets_s3
+        @app_secrets_s3 ||= Identity::Hostdata.app_secrets_s3(logger: @logger, s3_client: @s3_client)
+      end
+
+      def app_configuration_s3_path
+        "/%<env>s/#{app_configuration_path_component}/v1/application.yml"
+      end
+
+      def role_configuration_s3_path
+        return if role_configuration_filename.nil?
+        "/%<env>s/#{app_configuration_path_component}/v1/#{role_configuration_filename}"
+      end
+
+      def app_configuration_path_component
+        return 'idp' if Identity::Hostdata.instance_role == 'worker'
+        Identity::Hostdata.instance_role
```

Comment:
this will be like, `pki` in PKI app right?

Reply:
I'm pretty sure this actually returns `pivcac` in the PKI app. I'll hop on a host and check. The app.yml is actually under the name `pivcac`.

Reply:
It returns `pivcac` on the hosts which is what I was expecting.